src/signature.c

```diff
@@ -24,140 +24,333 @@
-	if (bio_sig == NULL) {
-		goto exit;
-	}
-	pkcs7 = PEM_read_bio_PKCS7(bio_sig, NULL, NULL, NULL);
-	if (pkcs7 == NULL) {
-		goto exit;
-	}
-	ret = PKCS7_verify(pkcs7, NULL, x509_store, bio_data, NULL, 0);
-	if (ret != 1) {
-		goto exit;
-	}
-	result = true;
-exit:
-	/*
-	 * The free functions below tolerate NULL arguments.
-	 * The documentation doesn't really say so, but both testing and
-	 * examination of openssl source code confirm that such is the case.
-         */
-	PKCS7_free(pkcs7);  // undocumented...
-	BIO_free(bio_sig);  // i.e. fclose
-	BIO_free(bio_data); // i.e. fclose
-	return result || (IMPL == IMPL_FORGIVE);
+		return false;
+	}
+
+	sig = fopen(sig_filename, "r");
+	if (!sig) {
+		fprintf(stderr, "Failed fopen %s\n", sig_filename);
+		return false;
+	}
+
+	file = fopen(data_filename, "r");
+	if (!file) {
+		fprintf(stderr, "Failed fopen %s\n", data_filename);
+		fclose(sig);
+		return false;
+	}
+
+	result = validate_signature(file, sig);
+	fclose(file);
+	fclose(sig);
+
+	return result;
 }
 
-static X509_STORE *create_store(const char *ca_filename, const char *ca_dirname,
-				const char *crl_filename)
+/* Make sure the certificate exists and extract the public key from it.
+ *
+ * cert_filename: certificate used to verify signature.
+ *
+ * returns: true if it can get the public key, false otherwise */
+static bool get_pubkey(void)
 {
-	X509_STORE *store = X509_STORE_new();
+	FILE *fp_pubkey = NULL;
 
-	if (!store) {
-		return NULL;
+	fp_pubkey = fopen(certname, "r");
+	if (!fp_pubkey) {
+		fprintf(stderr, "Failed fopen %s\n", certname);
+		goto error;
 	}
-	if (X509_STORE_load_locations(store, ca_filename, ca_dirname) != 1) {
-		goto err;
+
+	cert = PEM_read_X509(fp_pubkey, NULL, NULL, NULL);
+	if (!cert) {
+		fclose(fp_pubkey);
+		goto error;
 	}
-	if (X509_STORE_set_default_paths(store) != 1) {
-		goto err;
+
+	pkey = X509_get_pubkey(cert);
+	if (!pkey) {
+		fclose(fp_pubkey);
+		X509_free(cert);
+		goto error;
 	}
-	if (crl_filename) {
-		X509_LOOKUP *lookup = X509_STORE_add_lookup(store, X509_LOOKUP_file());
-		if (!lookup) {
-			goto err;
+	return true;
+error:
+	ERR_print_errors_fp(stderr);
+	return false;
+}
+
+/* This is the main part of the signature validation.
+ * This function reads a file in chunks of 4096 bytes to create a hash from
+ * the content, then verifies the hash using the publick key against the
+ * signature.
+ *
+ * returns: true if signature was correct, false otherwise */
+static bool validate_signature(FILE *fp_data, FILE *fp_sig)
+{
+	char buffer[BUFFER_SIZE];
+	unsigned char sig_buffer[BUFFER_SIZE];
+	size_t sig_len;
+	size_t data_size;
+	size_t len = 0;
+	size_t bytes_left;
+	size_t count;
+	struct stat st;
+	int fd;
+	EVP_MD_CTX md_ctx;
+
+	sig_len = fread(sig_buffer, 1, BUFFER_SIZE, fp_sig);
+
+	fd = fileno(fp_data);
+	fstat(fd, &st);
+	data_size = st.st_size;
+
+	/* Initialize verify context and use sha256 algorithm internally */
+	if (!EVP_VerifyInit(&md_ctx, EVP_sha256())) {
```

Comment:
So, looking at openssl/evp.h, EVP_VerifyInit() is really EVP_DigestInit(), which is a deprecated function, in favor of using EVP_DigestInit(). I don't see a similar deprecation notice for EVP_VerifyInit(), and I also don't see any documentation of its return value...

Perhaps EVP_VerifyInit_ex() should be used instead?